### PR TITLE
Develop semantic analysis

### PR DESCRIPTION
**Pull request by alkino**
- Use log system instead of only throw
- Add a new check: destructor blocks only in point process type mod file

### CODE REVIEW

**Comment by alkino** [comment]
@iomaganaris should we throw or only send message?

**Comment by iomaganaris** [comment]
> @iomaganaris should we throw or only send message?

Discussed offline. I suggested mimicking the behavior of `CodegenCompatibilityVisitor` (just printing warnings and returning 1 from the main function)

**Review by iomaganaris — COMMENTED** [review]
LGTM in general.
Small nitpicks and suggestions

File: test/unit/visitor/semantic_analysis.cpp [hunk 1]
```diff
@@ -59,8 +59,49 @@ SCENARIO("TABLE stmt", "[visitor][semantic_analysis]") {
...
+
+SCENARIO("Destructor block", "[visitor][semantic_analysis]") {
+    GIVEN("A point-process mod file, with a destructor") {
+        std::string nmodl_text = R"(
+            NEURON {
+                POINT_PROCESS test
+            }
+
+            DESTRUCTOR {
+            }
```
Comment (pramodk) [inline test/unit/visitor/semantic_analysis.cpp:76]: adapt this test to have `DESTRUCTOR` before`NEURON` block.

**Review by pramodk — CHANGES_REQUESTED** [review]

File: src/visitors/semantic_analysis_visitor.cpp [hunk 2]
```diff
@@ -1,31 +1,64 @@
...
+}
+
+void SemanticAnalysisVisitor::visit_destructor_block(const ast::DestructorBlock& node) {
+    /// <-- This code is for check 2
+    if (!is_point_process) {
+        logger->warn(
+            "SemanticAnalysisVisitor :: This mod file is not point process but contains a "
+            "destructor.");
+        check_fail = true;
     }
```
Comment (pramodk) [inline src/visitors/semantic_analysis_visitor.cpp:60]: If we have mod file containing `DESTRUCTOR` block before `NEURON` block:

```
DESTRUCTOR {

}

NEURON {
    THREADSAFE
    :POINT_PROCESS test
    SUFFIX test
}
```

above implementation wouldn't work. I would suggest to set `is_point_process` in `::check()` method (e.g. by looking up for `Suffix` node type in the whole AST)

**Review by pramodk — APPROVED** [review]
LGTM

**Pull request merged**